accumulate.go

```diff
@@ -31,66 +40,101 @@ namespace are stored.
 AccumulateKey. This is because the billing system can't handle multiple ObjectsUsers per namespace.
 */
-func accumulateBucketMetrics(ctx context.Context, date time.Time, cloudscaleClient *cloudscale.Client) (map[AccumulateKey]uint64, error) {
+func accumulateBucketMetrics(ctx context.Context, date time.Time, cloudscaleClient *cloudscale.Client, k8sclient client.Client) (map[AccumulateKey]uint64, error) {
 	bucketMetricsRequest := cloudscale.BucketMetricsRequest{Start: date, End: date}
 	bucketMetrics, err := cloudscaleClient.Metrics.GetBucketMetrics(ctx, &bucketMetricsRequest)
 	if err != nil {
 		return nil, err
 	}
 
+	nsTenants, err := fetchNamespaces(ctx, k8sclient, client.HasLabels{organizationLabel})
+	if err != nil {
+		return nil, err
+	}
+
+	buckets, err := fetchBuckets(ctx, k8sclient, client.HasLabels{namespaceLabel})
+	if err != nil {
+		return nil, err
+	}
+
 	accumulated := make(map[AccumulateKey]uint64)
 
 	for _, bucketMetricsData := range bucketMetrics.Data {
-		objectsUser, err := cloudscaleClient.ObjectsUsers.Get(ctx, bucketMetricsData.Subject.ObjectsUserID)
-		if err != nil || objectsUser == nil {
-			fmt.Fprintf(os.Stderr, "WARNING: Cannot sync bucket %s, objects user %s not found\n", bucketMetricsData.Subject.BucketName, bucketMetricsData.Subject.ObjectsUserID)
+		name := bucketMetricsData.Subject.BucketName
+		ns, ok := buckets[name]
+		if !ok {
+			fmt.Fprintf(os.Stderr, "WARNING: Cannot sync bucket, bucket resource %q not found\n", name)
 			continue
 		}
-		err = accumulateBucketMetricsForObjectsUser(accumulated, bucketMetricsData, objectsUser)
+		tenant, ok := nsTenants[ns]
+		if !ok {
+			fmt.Fprintf(os.Stderr, "WARNING: Cannot sync bucket, namespace %q not found in map\n", ns)
+			continue
+		}
+
+		err = accumulateBucketMetricsForObjectsUser(accumulated, bucketMetricsData, tenant, ns)
 		if err != nil {
-			fmt.Fprintf(os.Stderr, "WARNING: Cannot sync bucket %s: %v\n", bucketMetricsData.Subject.BucketName, err)
+			fmt.Fprintf(os.Stderr, "WARNING: Cannot sync bucket %s: %v\n", name, err)
 			continue
 		}
 	}
 
 	return accumulated, nil
 }
 
-func accumulateBucketMetricsForObjectsUser(accumulated map[AccumulateKey]uint64, bucketMetricsData cloudscale.BucketMetricsData, objectsUser *cloudscale.ObjectsUser) error {
-	if len(bucketMetricsData.TimeSeries) != 1 {
-		return fmt.Errorf("There must be exactly one metrics data point, found %d", len(bucketMetricsData.TimeSeries))
+func fetchBuckets(ctx context.Context, k8sclient client.Client, opts ...client.ListOption) (map[string]string, error) {
+	buckets := &cloudscalev1.BucketList{}
+	if err := k8sclient.List(ctx, buckets, opts...); err != nil {
+		return nil, fmt.Errorf("bucket list: %w", err)
 	}
 
-	tenantStr := objectsUser.Tags["tenant"]
-	if tenantStr == "" {
-		return fmt.Errorf("no tenant information found on objectsUser")
+	bucketNS := map[string]string{}
+	for _, b := range buckets.Items {
+		bucketNS[b.Spec.ForProvider.BucketName] = b.Labels[namespaceLabel]
```

Comment:
`spec.forProvider.bucketName` might actually be empty, in which case it defaults to the `metadata.name`.
Maybe use this function? https://github.com/vshn/provider-cloudscale/blob/master/apis/cloudscale/v1/bucket_types.go#L108

Reply:
makes sense, adjusted.